AstSemantics.md

```diff
@@ -96,107 +120,59 @@ following rules:
-variables do not have addresses and are not aliased in the globals or memory. Local
-variables have *Local types* and are initialized to the appropriate zero value for their
-type at the beginning of the function, except parameters which are initialized to the values
-of the arguments passed to the function.
+### Addressing
 
-  * `get_local`: read the current value of a local variable
-  * `set_local`: set the current value of a local variable
-
-The details of index space for local variables and their types will be further clarified,
-e.g. whether locals with type `int32` and `int64` must be contiguous and separate from 
-others, etc.
+Each linear memory access operation also has an address operand and an immediate
+integer offset attribute. The infinite-precision sum of the address operand's
+value with the offset attribute's value is called the *effective address*.
 
-## Control flow structures
+Effective addresses are interpreted as unsigned byte indices into the linear
+memory storage, starting at `0`. Linear memory accesses access memory starting
+at this byte and extending for the number of bytes implied by the memory
+attribute.
 
-WebAssembly offers basic structured control flow. All control flow structures
-are statements.
+Accesses to any bytes of linear memory beyond the `memory_size` are considered
+*out-of-bounds*. A module may optionally define that out-of-bounds includes
+small effective addresses close to `0`
+(see [discussion] (https://github.com/WebAssembly/design/issues/204)).
+The semantics of out-of-bounds accesses are discussed below.
 
-  * `block`: a fixed-length sequence of statements
-  * `if`: if statement
-  * `do_while`: do while statement, basically a loop with a conditional branch
-    (back to the top of the loop)
-  * `forever`: infinite loop statement (like `while (1)`), basically an
-    unconditional branch (back to the top of the loop)
-  * `continue`: continue to start of nested loop
-  * `break`: break to end from nested loop or block
-  * `return`: return zero or more values from this function
-  * `switch`: switch statement with fallthrough
+The use of infinite-precision in the effective address computation means that
+the addition of the offset to the address does is never wrapped, so if the
+address for an access is out-of-bounds, the effective address will always also
+be out-of-bounds. This is intended to simplify folding of offsets into complex
+address modes in hardware, and to simplify bounds checking optimizations.
 
-Break and continue statements can only target blocks or loops in which they are
-nested. This guarantees that all resulting control flow graphs are well-structured.
+In the MVP, address operands and offset attributes have type `int32`, and heap
+sizes are limited to 4 GiB. In the future, to support
+[>4GiB linear memory](FutureFeatures.md#heaps-bigger-than-4gib), there will be
+a mode in which they have type `int64`.
```

Comment:
Limited to 4GiB: link to Nondeterminism.md's description of "Environment-dependent resource limits may be exhausted".